Branch-SDK/src/io/branch/referral/indexing/BranchUniversalObject.java

```diff
@@ -0,0 +1,575 @@
+                    if (latestParam.has(Defines.Jsonkey.ContentImgUrl.getKey())) {
+                        branchUniversalObject.imageUrl_ = latestParam.getString(Defines.Jsonkey.ContentImgUrl.getKey());
+                    }
+                    if (latestParam.has(Defines.Jsonkey.ContentType.getKey())) {
+                        branchUniversalObject.type_ = latestParam.getString(Defines.Jsonkey.ContentType.getKey());
+                    }
+                    if (latestParam.has(Defines.Jsonkey.ContentExpiryTime.getKey())) {
+                        branchUniversalObject.expirationInMilliSec_ = latestParam.getLong(Defines.Jsonkey.ContentExpiryTime.getKey());
+                    }
+                    Iterator<String> keys = latestParam.keys();
+                    while (keys.hasNext()) {
+                        String key = keys.next();
+                        branchUniversalObject.addContentMetadata(key, latestParam.getString(key));
+                    }
+                }
+
+            } catch (Exception ignore) {
+            }
+        }
+        return branchUniversalObject;
+    }
+
+    //---------------------Marshaling and Unmarshaling----------//
+    @Override
+    public int describeContents() {
+        return 0;
+    }
+
+    public static final Parcelable.Creator CREATOR = new Parcelable.Creator() {
+        public BranchUniversalObject createFromParcel(Parcel in) {
+            return new BranchUniversalObject(in);
+        }
+
+        public BranchUniversalObject[] newArray(int size) {
+            return new BranchUniversalObject[size];
+        }
+    };
+
+    @Override
+    public void writeToParcel(Parcel dest, int flags) {
+        dest.writeString(canonicalIdentifier_);
+        dest.writeString(title_);
+        dest.writeString(description_);
+        dest.writeString(imageUrl_);
+        dest.writeString(type_);
+        dest.writeLong(expirationInMilliSec_);
+        dest.writeInt(indexMode_.ordinal());
+        dest.writeSerializable(keywords_);
+
+        int metaDataSize = metadata_.size();
+        dest.writeInt(metaDataSize);
+        for (HashMap.Entry<String, String> entry : metadata_.entrySet()) {
+            dest.writeString(entry.getKey());
+            dest.writeString(entry.getValue());
+        }
+    }
+
+    private BranchUniversalObject(Parcel in) {
+        this();
+
+        canonicalIdentifier_ = in.readString();
+        title_ = in.readString();
+        description_ = in.readString();
+        imageUrl_ = in.readString();
+        type_ = in.readString();
+        expirationInMilliSec_ = in.readLong();
+        indexMode_ = CONTENT_INDEX_MODE.values()[in.readInt()];
+        @SuppressWarnings("unchecked")
+        ArrayList<String> keywordsTemp = (ArrayList<String>) in.readSerializable();
+        keywords_.addAll(keywordsTemp);
+        int metadataSize = in.readInt();
+        for (int i = 0; i < metadataSize; i++) {
+            metadata_.put(in.readString(), in.readString());
```

Comment:
just to confirm, you're reading from the parcel and writing it back to the `HashMap` the same order it's written in? Parcel guarantees they'll be in order I assume.